pyccl/cls.py

```diff
@@ -308,27 +311,27 @@ def angular_cl(cosmo, cltracer1, cltracer2, ell,
     cosmo_in = cosmo
     cosmo = cosmo.cosmo
 
+    if p_of_k_a is not None:
+        if isinstance(p_of_k_a, Pk2D):
+            psp = p_of_k_a.psp
+        else:
+            raise ValueError("p_of_k_a must be either a "
+                             "pyccl.Pk2D object or None")
+    else:
+        psp = None
+
     # Access CCL_ClTracer objects
     clt1 = cltracer1.cltracer
     clt2 = cltracer2.cltracer
 
     status = 0
+    ell_use = np.atleast_1d(ell)
     # Return Cl values, according to whether ell is an array or not
-    if isinstance(ell, float) or isinstance(ell, int):
-        # Use single-value function
-        cl_one, status = lib.angular_cl_vec(
-            cosmo, clt1, clt2, l_limber, l_logstep,
-            l_linstep, [ell], 1, status)
-        cl = cl_one[0]
-    elif isinstance(ell, np.ndarray):
-        # Use vectorised function
-        cl, status = lib.angular_cl_vec(
-            cosmo, clt1, clt2, l_limber, l_logstep,
-            l_linstep, ell, ell.size, status)
-    else:
-        # Use vectorised function
-        cl, status = lib.angular_cl_vec(
-            cosmo, clt1, clt2, l_limber, l_logstep,
-            l_linstep, ell, len(ell), status)
+    cl, status = lib.angular_cl_vec(
+        cosmo, clt1, clt2, psp, l_limber, l_logstep,
+        l_linstep, ell_use, ell_use.size, status)
+    if ell_use.size==1:
```

Comment:
We can't test like this since a 1d array will always get coerced to a scalar. You can use `np.isscalar` and test `ell` directly.

Reply:
Fixed